cdap-data-fabric/src/main/java/co/cask/cdap/data/format/GrokRecordFormat.java

```diff
@@ -0,0 +1,179 @@
+    while (fields.hasNext()) {
+      Schema.Field field = fields.next();
+      Schema schema = field.getSchema();
+      // if we're not on the very last field, the field must be a simple type or a nullable simple type.
+      boolean isSimple = schema.getType().isSimpleType();
+      boolean isNullableSimple = schema.isNullableSimple();
+      if (!isSimple && !isNullableSimple) {
+        // if this is the very last field and a string array, it is valid. otherwise it is not.
+        if (fields.hasNext() || !isStringArray(schema)) {
+          throw new UnsupportedTypeException("Field " + field.getName() + " is of invalid type.");
+        }
+      }
+    }
+  }
+
+  // check that it's an array of strings or array of nullable strings. the array itself can also be nullable.
+  private boolean isStringArray(Schema schema) {
+    Schema arrSchema = schema.isNullable() ? schema.getNonNullable() : schema;
+    if (arrSchema.getType() == Schema.Type.ARRAY) {
+      Schema componentSchema = arrSchema.getComponentSchema();
+      if (componentSchema.isNullable()) {
+        return componentSchema.getNonNullable().getType() == Schema.Type.STRING;
+      } else {
+        return componentSchema.getType() == Schema.Type.STRING;
+      }
+    }
+    return false;
+  }
+
+  @Override
+  protected void configure(Map<String, String> settings) {
+    addPatterns(grok);
+
+    try {
+      this.pattern = determinePattern(settings);
+      grok.compile(pattern);
+    } catch (GrokException e) {
+      LOG.error("Failed to compile grok pattern '{}'", pattern, e);
+    }
+  }
+
+  protected void addPatterns(Grok grok) {
+    addPattern(grok, "cdap/grok/patterns/firewalls");
+    addPattern(grok, "cdap/grok/patterns/grok-patterns");
+    addPattern(grok, "cdap/grok/patterns/haproxy");
+    addPattern(grok, "cdap/grok/patterns/java");
+    addPattern(grok, "cdap/grok/patterns/junos");
+    addPattern(grok, "cdap/grok/patterns/linux-syslog");
+    addPattern(grok, "cdap/grok/patterns/mcollective");
+    addPattern(grok, "cdap/grok/patterns/mcollective-patterns");
+    addPattern(grok, "cdap/grok/patterns/mongodb");
+    addPattern(grok, "cdap/grok/patterns/nagios");
+    addPattern(grok, "cdap/grok/patterns/postgresql");
+    addPattern(grok, "cdap/grok/patterns/redis");
+    addPattern(grok, "cdap/grok/patterns/ruby");
+  }
+
+  protected String determinePattern(Map<String, String> settings) {
+    if (!settings.containsKey(PATTERN_SETTING)) {
+      return DEFAULT_PATTERN;
+    } else {
+      return settings.get(PATTERN_SETTING);
+    }
+  }
+
+  protected void addPattern(Grok grok, String resource) {
+    URL url = this.getClass().getClassLoader().getResource(resource);
+    if (url == null) {
+      LOG.error("Resource '{}' for grok pattern was not found", resource);
+      return;
+    }
+
+    try {
+      InputStream input = url.openStream();
+      try {
+        grok.addPatternFromReader(new InputStreamReader(input));
```

Comment:
Always specify charset when creating `Reader` from `InputStream`. Also, I suppose the pattern is relatively small, you can use `Resources.toString()` to read the content represented by the resource url and call `grok.addPattern`